File: WatchBox/Core/Catalog/TorrentStream.swift
//
//  TorrentStream.swift
//  SceneBox
//
//  Created by SpontaneousArray on 29.07.26.
//

import Foundation

nonisolated public struct TorrentStream: Identifiable, Sendable {
    public let id: String          // infoHash hex
    public let title: String       // full description line from Torrentio
    public let displayName: String // short label (source + resolution)
    public let infoHash: Data
    public let fileIndex: Int?     // which file in the torrent is the video
    public let trackers: [URL]
    public let seeders: Int?
    public let sizeText: String?
    public let resolution: String?
    public let url: URL?

    public var magnet: MagnetLink {
        MagnetLink(infoHash: infoHash, displayName: displayName, trackers: trackers)
    }

    public var isDebrid: Bool { url != nil }
}

File: lib/core/torrent/torrent_stream.dart
import '../models/media_type.dart';

class TorrentStream {
  const TorrentStream({
    required this.id,
    required this.title,
    required this.displayName,
    required this.infoHash,
    this.fileIndex,
    this.trackers = const [],
    this.seeders,
    this.sizeText,
    this.resolution,
    this.url,
  });

  final String id;
  final String title;
  final String displayName;
  final String infoHash;
  final int? fileIndex;
  final List<String> trackers;
  final int? seeders;
  final String? sizeText;
  final String? resolution;
  final Uri? url;

  bool get isDebrid => url != null;

  String get magnet {
    final params = <String>['xt=urn:btih:$infoHash'];
    if (displayName.isNotEmpty) {
      params.add('dn=${Uri.encodeComponent(displayName)}');
    }
    params.addAll(trackers.map((tracker) => 'tr=${Uri.encodeComponent(tracker)}'));
    return 'magnet:?${params.join('&')}';
  }
}
